doc/dev/schema.md

```diff
@@ -0,0 +1,193 @@
+
+``net.box`` schema reload works as follows. For each request (we use ``call``s
+to execute our procedures) from client (router) to server (storage) we
+receive a schema version together with response data. If schema versions mismatch,
+client reloads schema. The request is not retried. ``net.box`` reloads
+schema before returning ``call`` result to a user, so the next request will use
+updated schema. (See [tarantool/tarantool/6169](https://github.com/tarantool/tarantool/issues/6169)).
+``crud`` cannot implicitly rely on ``net.box`` reloads: ``crud`` requests
+use ``net.box`` spaces schema to build ``net.box`` ``call`` request data,
+so if something is not relevant anymore, a part of this request data need
+to be recalculated. ``crud`` uses connection ``reload_schema`` handle
+(``ping`` wrapper, see[PR#111 comment](https://github.com/tarantool/crud/pull/111#issuecomment-765811556))
+to reload schema explicitly when needed. ``crud`` reloads each replicaset
+schema. If there are several requests to reload info, the only one reload
+fiber is started and every request waits for its completion.
+
+### When schema is reloaded and operation is retried
+
+The basic logic is as follows: if something had failed and space schema
+mismatch could be the reason, reload the schema and retry. If it didn't
+help after ``N`` retries (now ``N`` is ``1``), pass the error to the
+user.
+
+Retry with reload is triggered if
+- space not found;
+- map flatten on `crud.*_object*` request has failed;
+- ``bucket_id`` calculation has failed due to any reason;
+- operation had failed on storage, hash check is on and hashes mismatch.
+
+Let's talk a bit more about the last one. To enable hash check, the user
+should pass `add_space_schema_hash` option to a request. This option is
+always enabled for `crud.*_object*` requests. If hashes mismatch, it
+means the router schema of this space is inconsistent with the storage
+space schema, so we reload it. For ``*_many`` methods, reload and retry
+happens only if all tuples had failed with hash mismatch; otherwise,
+errors are passed to a user.
+
+Retries with reload are processed with wrappers. Wrapper is a Lua
+function that wraps the call of some business logic Lua function (for
+example, a function that transforms `crud.insert_object` map to a tuple-like
+table or a function that performs insert request with this tuple-like table).
+Depending on return values, it decides to call some reload util and retry
+the function call or pass the error value to the user. Retries are counted
+per function it wraps, so in fact there could be more reloads than ``N``
+for a single request. For example, if a user calls `insert_object`, object
+flatten has failed (and has succeeded after reload) and then storage insert
+has failed and hashes mismatch, reload will happen twice and there will be
+two retries (though it's two different blocks of code will be retried
+independently). Or if a user calls `insert_object_many`, each tuple
```

Comment:
To be honest, I completely missed the point. I understand that when we have two-level retrying wrappers, we can get N*M retries at max. But here we don't say anything about sharding info wrappers. And we can't get more than N retries if the logic is like so:

```lua
local function pure_crud_request()
    <...>
end

local function crud_request_with_schema_reload()
    while i = 1, N do
        pure_crud_request()
        if <..ok..> then return <..result..>
        if <..schema should be reloaded..> then continue end
        return <..error..>
    end
    return <..last error..>
end
```

Maybe it worth to illustrate the words with a pseudocode?

Reply:
Actually it is not as you expected. Reworked with code example, hope it will be enough to understand.

Reply:
I got it now, thanks!